Add `SplitElem`

T0mstone opened the pull request:
This is basically the same as #1775 but distilled to its essence.

What `instance-block` from #1775 did with a function is now idiomatically delegated to a show rule.

Feel free to review and accept or reject. I'm just opening this for completeness.

It still has the same issue/feature that the individual parts of broken content have a fixed layout, which can cause weirdness; Sadly, I failed to come up with a better solution.

LuxxxLucy commented:
hi would you add some test cases? (Either add and explain here in the PR page or committed in the tests/typ/ref)

I think we will need at least one working example and then maybe a "weird" case as you mentioned 

> It still has the same issue/feature that the individual parts of broken content have a fixed layout, which can cause weirdness; Sadly, I failed to come up with a better solution.

T0mstone commented:
Here's an example:
```typst
#set page("a6")

// normal example
#{
	show split.item: it => {
		let col = if it.index == 0 {
			red
		} else if it.index == it.count - 1 {
			blue
		} else {
			green
		}
		rect(stroke: col, inset: 2pt, it)
	}

	columns(3, split(lorem(100)))
}

#pagebreak()

// weirdness
#{
	show split.item: it => {
		let col = if it.index == 0 {
			red
		} else if it.index == it.count - 1 {
			blue
		} else {
			green
		}
		rect(stroke: col, inset: 2pt)[My Box\ #it]
	}

	columns(3, split(lorem(100)))
}
```
![normal-example](https://github.com/typst/typst/assets/39707032/c9699a78-282d-4f8d-8002-d34eba5a6a7e)
![weirdness](https://github.com/typst/typst/assets/39707032/f8aa58da-5efd-439c-bb85-9c3dc36f5f6e)

LuxxxLucy commented:
Hi, I think the first result is neat and I can see why the split could be useful. But I felt the same *Cursedness* you mentioned in #1775 that is preventing me to like the implementation.  I am not convinced by the idea of placeholder. 

Is there any way to avoid the use of the replace-placeholder approach and instead trigger the system to re-layout? To me, the relayout is still a must-do for a typesetting system. 

Need input from @laurmaedje To me while the result seems neat (the first one) the use of placeholder and then replacing the frame approach (same in this PR and in #1775) feels somehow not very preferable

laurmaedje commented:
The whole layout engine is in a tricky spot. Medium-term, we need to put the whole layout on a more solid foundation. For that we need a good set of primitives both for normal things and also more cetz-like stuff for relative positioning. I am unsure how to proceed with layout PRs in the meantime.

laurmaedje commented:
I think this isn't something which should be merged as-is, both because of the weirdness that it can't affected the item layout and because a the current general lack of direction for the layout engine. I think we should revisit the general idea of this when the layout engine rework is shifted into focus.
